Make it possible to set -_topContentInset with an immediate push to the web process

smfr opened the pull request:
#### 07f28d8dc23bc8916423610af77e4d6c2f69b15a
<pre>
Make it possible to set -_topContentInset with an immediate push to the web process
<a href="https://bugs.webkit.org/show_bug.cgi?id=282857">https://bugs.webkit.org/show_bug.cgi?id=282857</a>
<a href="https://rdar.apple.com/139542356">rdar://139542356</a>

Reviewed by Aditya Keerthi.

152497@main added some asynchrony to setting _topContentInset, delaying the push to
the web process by a runloop turn by default. This causes problems for clients that
attempt to change the WKWebView frame and _topContentInset at the same time, and expect
flicker-free rendering.

This change doesn&apos;t fix the fundamental issue that changing the view&apos;s frame and
_topContentInset isn&apos;t synchronized (because the visible impact of changing
_topContentInset require a round-trip via the web process), but it does allow the client
to avoid the `RunLoop::main().dispatch()` asynchrony in `scheduleSetTopContentInsetDispatch`
via a new `-[WKWebView _setTopContentInset:immediate:]` SPI.

In addition, if the view size is changed, we ensure that we push a pending topContentInset
change to the web process in `RemoteLayerTreeDrawingAreaProxy::sendUpdateGeometry()`. This
requires moving the storage of `m_pendingTopContentInset` and related data and functions
from WebViewImpl to WebPageProxy. We use a PageClient callback to compute the
automatic topContentInset.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _setTopContentInset:]):
(-[WKWebView _setTopContentInset:immediate:]):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::computeAutomaticTopContentInset):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::sendUpdateGeometry):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::computeAutomaticTopContentInset):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::setAutomaticallyAdjustsContentInsets):
(WebKit::WebPageProxy::updateContentInsetsIfAutomatic):
(WebKit::WebPageProxy::setTopContentInsetAsync):
(WebKit::WebPageProxy::pendingOrActualTopContentInset const):
(WebKit::WebPageProxy::scheduleSetTopContentInsetDispatch):
(WebKit::WebPageProxy::dispatchSetTopContentInset):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setAutomaticallyAdjustsContentInsets):
(WebKit::WebViewImpl::automaticallyAdjustsContentInsets const):
(WebKit::WebViewImpl::updateContentInsetsIfAutomatic):
(WebKit::WebViewImpl::topContentInset const):
(WebKit::WebViewImpl::setTopContentInset):
(WebKit::WebViewImpl::flushPendingTopContentInset):
(WebKit::WebViewImpl::endDeferringViewInWindowChanges):
(WebKit::WebViewImpl::endDeferringViewInWindowChangesSync):
(WebKit::WebViewImpl::prepareForMoveToWindow):
(WebKit::WebViewImpl::scheduleSetTopContentInsetDispatch): Deleted.
(WebKit::WebViewImpl::dispatchSetTopContentInset): Deleted.

Canonical link: <a href="https://commits.webkit.org/286379@main">https://commits.webkit.org/286379@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d47777d3cde3e711dbaaaad714518af1b0cf2630

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75848 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54877 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28745 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80344 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27114 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64019 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3171 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59472 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17633 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78915 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49343 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65142 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39832 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46742 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25441 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67858 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22963 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81809 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3217 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2018 "Found 1 new test failure: imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67701 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3371 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65109 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67007 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16702 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10950 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9077 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3167 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5973 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3188 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4126 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3195 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
```diff
@@ -1450,14 +1450,21 @@ - (void)_setDrawsBackground:(BOOL)drawsBackground
 - (void)_setTopContentInset:(CGFloat)contentInset
 {
-    return _impl->setTopContentInset(contentInset);
+    _impl->setTopContentInset(contentInset);
```
Maybe use `[self _setTopContentInset:contentInset immediate:NO]` here?

pxlcoder commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -1023,6 +1023,17 @@ static inline NSCorrectionResponse toCorrectionResponse(AutocorrectionResponse r
+std::optional<float> PageClientImpl::computeAutomaticTopContentInset()
+{
+    RetainPtr window = [m_view window];
+    if ((window.get().styleMask & NSWindowStyleMaskFullSizeContentView) && !window.get().titlebarAppearsTransparent && ![m_view enclosingScrollView]) {
```
```suggestion
    if (([window styleMask] & NSWindowStyleMaskFullSizeContentView) && ![window titlebarAppearsTransparent] && ![m_view enclosingScrollView]) {
```

pxlcoder commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -1023,6 +1023,17 @@ static inline NSCorrectionResponse toCorrectionResponse(AutocorrectionResponse r
+{
+    RetainPtr window = [m_view window];
+    if ((window.get().styleMask & NSWindowStyleMaskFullSizeContentView) && !window.get().titlebarAppearsTransparent && ![m_view enclosingScrollView]) {
+        NSRect contentLayoutRectInWebViewCoordinates = [m_view convertRect:window.get().contentLayoutRect fromView:nil];
```
```suggestion
        NSRect contentLayoutRectInWebViewCoordinates = [m_view convertRect:[window contentLayoutRect] fromView:nil];
```

pxlcoder commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -390,6 +390,69 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+
+    m_didScheduleSetTopContentInsetDispatch = true;
+
+    RunLoop::main().dispatch([weakThis = WeakPtr { *this }] {
```
```suggestion
    callOnMainRunLoop([weakThis = WeakPtr { *this }] {
```

pxlcoder commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -390,6 +390,69 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+    if (!m_didScheduleSetTopContentInsetDispatch)
+        return;
+
+    m_didScheduleSetTopContentInsetDispatch = false;
```
Could use `std::exchange` here.

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -1825,68 +1825,34 @@ static bool isInRecoveryOS()
-        return;
-
-    m_didScheduleSetTopContentInsetDispatch = true;
+    WTF_ALWAYS_LOG("WebViewImpl::setTopContentInset " << (float)contentInset);
```
Unintentionally included?

smfr committed:
Make it possible to set -_topContentInset with an immediate push to the web process
https://bugs.webkit.org/show_bug.cgi?id=282857
rdar://139542356

Reviewed by Aditya Keerthi.

152497@main added some asynchrony to setting _topContentInset, delaying the push to
the web process by a runloop turn by default. This causes problems for clients that
attempt to change the WKWebView frame and _topContentInset at the same time, and expect
flicker-free rendering.

This change doesn't fix the fundamental issue that changing the view's frame and
_topContentInset isn't synchronized (because the visible impact of changing
_topContentInset require a round-trip via the web process), but it does allow the client
to avoid the `RunLoop::main().dispatch()` asynchrony in `scheduleSetTopContentInsetDispatch`
via a new `-[WKWebView _setTopContentInset:immediate:]` SPI.

In addition, if the view size is changed, we ensure that we push a pending topContentInset
change to the web process in `RemoteLayerTreeDrawingAreaProxy::sendUpdateGeometry()`. This
requires moving the storage of `m_pendingTopContentInset` and related data and functions
from WebViewImpl to WebPageProxy. We use a PageClient callback to compute the
automatic topContentInset.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _setTopContentInset:]):
(-[WKWebView _setTopContentInset:immediate:]):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::computeAutomaticTopContentInset):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::sendUpdateGeometry):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::computeAutomaticTopContentInset):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::setAutomaticallyAdjustsContentInsets):
(WebKit::WebPageProxy::updateContentInsetsIfAutomatic):
(WebKit::WebPageProxy::setTopContentInsetAsync):
(WebKit::WebPageProxy::pendingOrActualTopContentInset const):
(WebKit::WebPageProxy::scheduleSetTopContentInsetDispatch):
(WebKit::WebPageProxy::dispatchSetTopContentInset):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setAutomaticallyAdjustsContentInsets):
(WebKit::WebViewImpl::automaticallyAdjustsContentInsets const):
(WebKit::WebViewImpl::updateContentInsetsIfAutomatic):
(WebKit::WebViewImpl::topContentInset const):
(WebKit::WebViewImpl::setTopContentInset):
(WebKit::WebViewImpl::flushPendingTopContentInset):
(WebKit::WebViewImpl::endDeferringViewInWindowChanges):
(WebKit::WebViewImpl::endDeferringViewInWindowChangesSync):
(WebKit::WebViewImpl::prepareForMoveToWindow):
(WebKit::WebViewImpl::scheduleSetTopContentInsetDispatch): Deleted.
(WebKit::WebViewImpl::dispatchSetTopContentInset): Deleted.

Canonical link: https://commits.webkit.org/286379@main
